src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -259,6 +259,84 @@ public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancella
             }
         }
 
+        /// <inheritdoc />
+        public override async Task CopyToAsync(PipeWriter destination, CancellationToken cancellationToken = default)
+        {
+            ThrowIfCompleted();
+
+            // PERF: store InternalTokenSource locally to avoid querying it twice (which acquires a lock)
+            CancellationTokenSource tokenSource = InternalTokenSource;
+            if (TryReadInternal(tokenSource, out ReadResult readResult))
+            {
+                await WriteAsyncInternal(
+                    destination,
+                    (destination, memory, cancellationToken) => destination.WriteAsync(memory, cancellationToken),
+                    readResult,
+                    cancellationToken).ConfigureAwait(false);
+            }
+
+            if (_isStreamCompleted) return;
+
+            CancellationTokenRegistration reg = default;
+            if (cancellationToken.CanBeCanceled)
+            {
+                reg = cancellationToken.UnsafeRegister(state => ((StreamPipeReader)state!).Cancel(), this);
+            }
```

Comment:
I think this code should just manually write the existing buffers into the destination instead of calling these APIs.

Reply:
Also fixed. Should we avoid the code duplication here? If yes, should we apply a similar factorization as in `PipeReader` between the `Stream` and `PipeReader` implementations?

Reply:
Yes but let me see what it looks like before that refactoring.